File: Module_1/performance/print_primes.py
"""
A simple python program to test CPU performance by
generating a series of prime numbers
"""

from threading import Thread, current_thread
from time import time
import sys

NUM_SERIES = 10_000_000
NUM_THREADS = 4


def is_prime(number):
    """Returns True if 'number' is prime"""
    limit = int(number ** 0.5) + 1
    for d in range(2, limit):
        if number % d == 0:
            return False
    return True


def print_primes(start, stop):
    """Store prime numbers from start to stop into primes
    list passed as arguments
    """
    tname = current_thread().name

    while start < stop:
        if is_prime(start):
            print(f"{tname}: {start}")
        start += 1


if __name__ == '__main__':

    threads = []
    chunk_size = NUM_SERIES // NUM_THREADS

    start_time = time()
    for i in range(NUM_THREADS):
        start = i * chunk_size
        stop = (i * chunk_size) + chunk_size
        t = Thread(target=print_primes, args=(start, stop))
        t.start()
        threads.append(t)
        print(f"Created {t.name} with {start=}, {stop=}", file=sys.stderr)

    for t in threads:
        t.join()
    duration = time() - start_time

    print(f"{NUM_THREADS} Threads took {duration:.6f} seconds to complete",
          file=sys.stderr)

File: Module_1/performance/print_primes.rs
// A simple Rust program to test CPU performance by
// generating a series of prime numbers

use std::thread;
use std::time::Instant;

const NUM_SERIES: u64 = 10_000_000;
const NUM_THREADS: u64 = 4;

fn is_prime(number: u64) -> bool {
    let limit = (number as f64).sqrt() as u64 + 1;
    for d in 2..limit {
        if number % d == 0 {
            return false;
        }
    }
    return true;
}

fn print_primes(start: u64, stop: u64, thread_name: String) {
    let mut current = start;
    while current < stop {
        if is_prime(current) {
            println!("{}: {}", thread_name, current);
        }
        current += 1;
    }
}

fn main() {
    let mut threads = vec![];
    let chunk_size: u64 = NUM_SERIES / NUM_THREADS;

    let start_time = Instant::now();
    for i in 0..NUM_THREADS {
        let start: u64 = i * chunk_size;
        let stop: u64 = start + chunk_size;
        let thread_name = format!("Thread-{}", i + 1);

        eprintln!("Created thread {} with start={}, stop={}", thread_name, start, stop);

        let t = thread::spawn(move || {
            print_primes(start, stop, thread_name);
        });

        threads.push(t);
    }


    for t in threads {
        t.join().unwrap();
    }
    let duration = start_time.elapsed();
    eprintln!("{} Threads took {:.6} seconds to complete", NUM_THREADS, duration.as_secs_f64());
}
